lib/samly/helper.ex

```diff
@@ -68,12 +68,19 @@ defmodule Samly.Helper do
   end
 
   def decode_idp_auth_resp(sp, saml_encoding, saml_response) do
-    with {:ok, xml_frag} <- decode_saml_payload(saml_encoding, saml_response),
-         {:ok, assertion_rec} <- :esaml_sp.validate_assertion(xml_frag, sp) do
-      {:ok, Assertion.from_rec(assertion_rec)}
-    else
-      {:error, reason} -> {:error, reason}
-      error -> {:error, {:invalid_request, "#{inspect(error)}"}}
+    case decode_saml_payload(saml_encoding, saml_response) do
+      {:ok, xml_frag} ->
+        case :esaml_sp.validate_assertion(xml_frag, sp) do
+          {:ok, assertion_rec} ->
+            {:ok, Assertion.from_rec(assertion_rec)}
+
+          {:error, reason} ->
+            Trace.handle("SAML assertion validation failed", error: reason, xml: xml_frag)
+            {:error, reason}
+        end
+
+      error ->
+        {:error, {:invalid_request, "#{inspect(error)}"}}
```

Comment:
I can see we are going to have ongoing disagreement about `with` vs nested `case` :)

Reply:
The issue here wasn't that I didn't like it, it was that I needed to report the value of xml_frag in the error case.